### PR TITLE
Performance fix for point in polygon

### PR DESCRIPTION
**Pull request by b4l**
- [x] I agree to follow the project's [code of conduct](https://github.com/georust/geo/blob/main/CODE_OF_CONDUCT.md).
- [ ] I added an entry to `CHANGES.md` if knowledge of this change could be valuable to users.
---

### CODE REVIEW

File: geo/src/algorithm/coordinate_position.rs [hunk 1]
```diff
@@ -239,11 +239,6 @@ where
             return;
         }
 
-        // Ok to `unwrap` since we know `self` is non-empty, so bounding-rect is non-null
```
Comment (michaelkirk) [inline geo/src/algorithm/coordinate_position.rs:242]: Since bbox checks are typically pretty quick, this was a short-circuit for when the point is "far" outside the ring. 

I honestly expected to see regressions for that case, so I just now wrote an additional bench that tests points far from their ring:
https://github.com/georust/geo/commit/efac869b7b7430bb27bb91922bacdf4bcfaa1e21

...but I was wrong! You've sped up the winding number logic so much that even computing a bbox is slower.

File: geo/src/algorithm/coordinate_position.rs [hunk 2]
```diff
@@ -371,21 +366,30 @@ where
     // See: https://en.wikipedia.org/wiki/Point_in_polygon#Winding_number_algorithm
     let mut winding_number = 0;
     for line in linestring.lines() {
+        // Edge Crossing Rules:
+        //   1. an upward edge includes its starting endpoint, and excludes its final endpoint;
+        //   2. a downward edge excludes its starting endpoint, and includes its final endpoint;
+        //   3. horizontal edges are excluded
+        //   4. the edge-ray intersection point must be strictly right of the point P.
```
Comment (michaelkirk) [inline geo/src/algorithm/coordinate_position.rs:373]: What is "the point P" referring to in these notes?

File: geo/src/algorithm/coordinate_position.rs [hunk 3]
```diff
@@ -371,21 +366,30 @@ where
...
         if line.start.y <= coord.y {
-            let o = T::Ker::orient2d(line.start, line.end, coord);
-            if o == Orientation::Collinear && point_in_rect(coord, line.start, line.end) {
-                return CoordPos::OnBoundary;
-            }
-            if line.end.y > coord.y && o == Orientation::CounterClockwise {
-                winding_number += 1;
-            }
+            if line.end.y >= coord.y {
```
Comment (michaelkirk) [inline geo/src/algorithm/coordinate_position.rs:375]: I'm trying to understand why these changes make things faster. 

With this new check `line.end.y >= coord.y`, we ensure we only call orient2d if a horizontal ray passing through `coord` would also pass through `line`. Checking `line.end.y >= coord.y` is much cheaper than orient2d, so we gain speed there.

But then, shouldn't we have a similar check (`if line.start.y >= coord.y {`) for the downward edge case below?

I confess that I tried it, but saw somewhere between small and no gains. I don't understand why though.

**Review by michaelkirk — APPROVED** [review]
Our `contains` benchmarks show some really great improvements from this!

```
point in simple polygon time:   [7.0496 ns 7.0730 ns 7.1007 ns]                                                                                                                            
                        change: [-55.196% -55.112% -55.020%] (p = 0.00 < 0.05)                                                                                                             
                        Performance has improved.                                                                                                                                          

point outside simple polygon
                        time:   [4.3600 ns 4.3647 ns 4.3701 ns]
                        change: [-23.790% -23.631% -23.451%] (p = 0.00 < 0.05)

point inside complex polygon
                        time:   [870.00 ns 871.77 ns 873.88 ns]
                        change: [-86.956% -86.892% -86.833%] (p = 0.00 < 0.05)
                        Performance has improved.

point outside complex polygon
                        time:   [876.98 ns 878.42 ns 880.08 ns]
                        change: [-85.146% -85.114% -85.082%] (p = 0.00 < 0.05)
                        Performance has improved.

line across complex polygon
                        time:   [331.40 µs 332.98 µs 334.75 µs]
                        change: [-4.7109% -3.7784% -2.8551%] (p = 0.00 < 0.05)
                        Performance has improved.

complex polygon contains polygon
                        time:   [509.71 µs 511.27 µs 512.83 µs]
                        change: [-4.8577% -4.3909% -3.9272%] (p = 0.00 < 0.05)
                        Performance has improved.
```

I confess I'm not super confident in my review of the algorithm. Some testing around edge cases, like if the ray is passing through an internal or external vertex would be nice, but I'm biased towards merging it even without that.

I'll leave it open another day or two in case anyone else wants to take a peak.

Thanks for the pr!

**Comment by michaelkirk** [comment]
> Some testing around edge cases, like if the ray is passing through an internal or external vertex would be nice, but I'm biased towards merging it even without that.

Actually! It looks like we already this: see `point_in_polygon_with_ray_passing_through_a_vertex_test` and neighbors.

**Comment by urschrei** [comment]
> > Some testing around edge cases, like if the ray is passing through an internal or external vertex would be nice, but I'm biased towards merging it even without that.
> 
> 
> 
> Actually! It looks like we already this: see `point_in_polygon_with_ray_passing_through_a_vertex_test` and neighbors.
> 
> 
> 
> 

I love past, diligent [whoever contrived that test]

**Comment by michaelkirk** [comment]
Doing a little git archeology... @bjornharrtell is responsible for most of the current implementation, introduced in https://github.com/georust/geo/pull/972.

Maybe @bjornharrtell would like to review?

And the good tests were from @amatissart as part of https://github.com/georust/geo/pull/526

**Comment by bjornharrtell** [comment]
Nice work 🙂

File: geo/src/algorithm/coordinate_position.rs (hunk 2 above)
Comment (b4l) [reply geo/src/algorithm/coordinate_position.rs:373]: The point to test for intersection/location/inclusion.

**Comment by bjornharrtell** [comment]
Had a second check and nice catch about the orientation checks. I don't remember why I ended up using those as they were not in the original code that I ported...

File: geo/src/algorithm/coordinate_position.rs (hunk 3 above)
Comment (b4l) [reply geo/src/algorithm/coordinate_position.rs:375]: Its basically this, that for all edges with starting y coord below the ray, the orientation was calculated before checking for the line ending above the line.

> But then, shouldn't we have a similar check (`if line.start.y >= coord.y {`) for the downward edge case below?

This is implicitly given by the test in the if clause `line.start.y <= coord.y` which gives us `line.start.y > coord.y` in the else clause what matches with the edgecrossing rule 2. a downward edge excludes its starting endpoint.

**Comment by b4l** [comment]
Thanks for the reviews!

Honestly the tests were what gave me confidence to open the pr in the first place.

I think the only open issue is whether to put back in the proceeding bounding rect intersection test. No preference here.

**Comment by michaelkirk** [comment]
> I think the only open issue is whether to put back in the proceeding bounding rect intersection test. No preference here.

My best guess is to leave it out, since it seemed to only slow things down in all of our benchmarks. 

How'd you feel about including this bench though? I think it's an interesting case if we revisit this code in the future: https://github.com/georust/geo/commit/efac869b7b7430bb27bb91922bacdf4bcfaa1e21

File: geo/src/algorithm/coordinate_position.rs (hunk 1 above)
Comment (michaelkirk) [reply geo/src/algorithm/coordinate_position.rs:242]: That seems well reasoned to me. Thank you for explaining.

Comment (michaelkirk) [reply geo/src/algorithm/coordinate_position.rs:242]: Just to make sure I'm understanding... this is the kind of case where we might expect a regression right?

Where the point (green) is horizontal (blue) to an inordinate number of edges in the polygon (red).

image

(forgive my graphic art skills)

File: geo/src/algorithm/coordinate_position.rs (hunk 2 above)
Comment (michaelkirk) [reply geo/src/algorithm/coordinate_position.rs:373]: So can we call it "`coord`" to match with the code it's documenting - we don't use `P` anywhere else in the docs or code.

File: geo/src/algorithm/coordinate_position.rs (hunk 1 above)
Comment (b4l) [reply geo/src/algorithm/coordinate_position.rs:242]: yes, exactly! we could prevent this with a bounding check, though not when the point is inside the bounds. for the latter some randomized ray direction shenanigans could help.

File: geo/src/algorithm/coordinate_position.rs (hunk 2 above)
Comment (b4l) [reply geo/src/algorithm/coordinate_position.rs]: yes, done!

**Comment by b4l** [comment]
> How'd you feel about including this bench though? I think it's an interesting case if we revisit this code in the future: [efac869](https://github.com/georust/geo/commit/efac869b7b7430bb27bb91922bacdf4bcfaa1e21)

Sure, if it helps. Performance wise your nice comb would be even better including a version with the point inside the bounds.

File: geo/src/algorithm/coordinate_position.rs (hunk 3 above)
Comment (michaelkirk) [reply geo/src/algorithm/coordinate_position.rs:375]: oh duh, `line.start.y >= coord.y` is always true at that point. Thank you. =)

**Comment by michaelkirk** [comment]
> Sure, if it helps. Performance wise your nice comb would be even better including a version with the point inside the bounds.

Sure, I'll throw one together right now, standby.

**Comment by michaelkirk** [comment]
Ok done! https://github.com/georust/geo/compare/coord-pos...mkirk/coord-pos?expand=1

If it looks good to you, merge it into this branch, and then we can merge the whole shebang into main.

**Comment by michaelkirk** [comment]
btw, here's the bench diff for that new case:

```
point horizontal to comb teeth aka bart's haircut
                        time:   [16.153 ns 16.169 ns 16.189 ns]
                        change: [+156.71% +157.26% +157.85%] (p = 0.00 < 0.05)
                        Performance has regressed.
```

I think that's fine given the pathological nature of the input.

**Comment by b4l** [comment]
@michaelkirk thanks! I hope I did this right with the merge.

**Review by michaelkirk — COMMENTED** [review]
bors r+
